### PR TITLE
[APPC-1691] Check for purchases made by guest wallet

### PR DESCRIPTION
**Pull request by JDSM01**
**What does this PR do?**

Checks for purchases made by the guest wallet, with or without wallet installed
Consumes unconsumed purchases made by the guest wallet, with or without wallet installed

**Database changed?**

   Yes | No

**Where should the reviewer start?**

AppCoinsStubHelper.java
AppCoinsBillingWrapper.java

**How should this be manually tested?**

1- On toolbox, press the query purchases without wallet and check if it returns an unconsumed purchase (you have to use debug, since toolbox is unable to verify the purchase)
2- Install the AppCoinsWallet
3- Do step 1 again
4- Do a purchase with the Wallet
5- Do step 1 again, and check if the list has 2 items

**What are the relevant tickets?**

https://aptoide.atlassian.net/browse/APPC-1691

**Code Review Checklist**

- [ ] Architecture
- [ ] Documentation on public interfaces
- [ ] Database changed?
- [ ] If yes - Database migration?
- [ ] Remove comments & unused code & forgotten testing Logs
- [ ] Codestyle
- [ ] Functional QA tests pass

### CODE REVIEW

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java [hunk 1]
```diff
@@ -6,6 +6,7 @@
 
 public class SharedPreferencesRepository {
 
+  public static final int TTL_IN_SECONDS = 86400 * 30; //86400 = 24h
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/SharedPreferencesRepository.java:9]: this variable is named as seconds and then used as days

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AppcoinsBillingWrapper.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      GuestPurchasesInteract guestPurchasesInteract = new GuestPurchasesInteract(billingRepository);
+      CountDownLatch countDownLatch = new CountDownLatch(1);
+      guestPurchasesInteract.mapGuestPurchases(bundle, walletId, packageName, type, countDownLatch,
+          idsList, skuList, dataList, signatureDataList);
+      waitForPurchases(countDownLatch);
+    }
+    return bundle;
+  }
+
+  @Override public int consumePurchase(int apiVersion, String packageName, String purchaseToken)
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AppcoinsBillingWrapper.java:66]: can we make this method more clear?maybe get both responses and calculate the response to be returned at the end of the method?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AppcoinsBillingWrapper.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    } catch (InterruptedException e) {
+      e.printStackTrace();
+    }
+  }
+
+  private int consumeGuestPurchase(String packageName, String purchaseToken) {
+    int[] responseCode = new int[1]; //Generic error
+    CountDownLatch countDownLatch = new CountDownLatch(1);
+    BillingRepository billingRepository =
+        new BillingRepository(new BdsService(BuildConfig.HOST_WS, 30000));
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AppcoinsBillingWrapper.java]: why don't we use the generic constant for timeout? `BdsService.TIME_OUT_IN_MILLIS`

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AppcoinsBillingWrapper.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  private void waitForPurchases(CountDownLatch countDownLatch) {
+    try {
+      countDownLatch.await(timeoutInMillis, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      e.printStackTrace();
+    }
+  }
+
+  private int consumeGuestPurchase(String packageName, String purchaseToken) {
+    int[] responseCode = new int[1]; //Generic error
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AppcoinsBillingWrapper.java]: using this array turns the code hard to understand, is there anyway to change it?

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AppcoinsBillingStubHelper.java [hunk 5]
```diff
@@ -144,40 +153,87 @@ public Bundle getSkuDetails(final int apiVersion, final String packageName, fina
...
 
+  private int consumeGuestPurchase(String walletId, String packageName, String purchaseToken) {
+    int[] responseCode = new int[1]; //Generic error
+    CountDownLatch countDownLatch = new CountDownLatch(1);
+    BillingRepository billingRepository =
+        new BillingRepository(new BdsService(BuildConfig.HOST_WS, 30000));
+    GuestPurchasesInteract guestPurchaseInteract = new GuestPurchasesInteract(billingRepository);
+
+    guestPurchaseInteract.consumeGuestPurchase(walletId, packageName, purchaseToken, responseCode,
+        countDownLatch);
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AppcoinsBillingStubHelper.java]: i think that you shouldn't send this countDownLatch here, it's not Interactor's responsability to deal this the syncronization, if it is, it should handle it by itself

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java [hunk 6]
```diff
@@ -96,9 +96,13 @@ private void navigateTo(Fragment fragment) {
         .commit();
   }
 
-  @Override public void close() {
+  @Override public void close(boolean withError) {
     Bundle bundle = new Bundle();
-    bundle.putInt(RESPONSE_CODE, 1); //CANCEL
+    if (withError) {
+      bundle.putInt(RESPONSE_CODE, 6); //ERROR
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java]: why don't you use the constants?

**Review by Trinkes — CHANGES_REQUESTED** [review]

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AppcoinsBillingWrapper.java (hunk 3 above)
Comment (JDSM01) [reply android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/helpers/AppcoinsBillingWrapper.java]: I added it in another PR and forgot to use it here

File: android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java [hunk 7]
```diff
@@ -42,7 +42,7 @@
 
     //This log is necessary for the automatic test that validates the wallet installation dialog
     Log.d("InstallDialog", "com.appcoins.sdk.billing.helpers.InstallDialogActivity started");
-
+    Log.d("TAG123", "HERE");
```
Comment (Trinkes) [inline android-appcoins-billing/src/main/java/com/appcoins/sdk/billing/payasguest/IabActivity.java]: is this log supposed to be here?

**Review by Trinkes — CHANGES_REQUESTED** [review]

**Review by Trinkes — APPROVED** [review]

**Pull request merged**